### DISCUSSION TITLE
Eloquent Custom Casts fire database queries on each save()

### DISCUSSION THREAD

**Discussion by aymericauberger** [laravel/framework#32808] [General]
- Laravel Version: 7.x
- PHP Version: 7.4.3

### Description:
When using an Eloquent custom cast on a model, if the value has been retrieved but not changed and we hit the `save()` method, a database query will be fired to re-insert the value and update the timestamps.

I believe this is because the `originalIsEquivalent` method on the `Illuminate\Database\Eloquent\Concerns\HasAttributes` trait does not support custom casts and will always return false.

### Steps To Reproduce:
Assuming we have a custom cast on the `address` property of the `User` model:
```
$user = User::first();
if ($user->address) {
  // Nothing changed...
}
$user->save();
```

### Possible solution:
We could add a method to be implemented on the cast class, for instance `isEquivalent($attribute, $original)`, to determine if the value was changed or not.

**Comment by taylorotwell** [comment]
I'm going to convert this to a discussion since it is a bit of a feature request atm instead of a "bug".

**Comment by aymericauberger** [comment]
Thanks! I wasn't sure where to put this, as it isn't a bug but I was expecting it would just work like the other cast types.
I can try to submit a PR, unless someone thinks of a better solution than the one I suggested?
